rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java

```diff
@@ -0,0 +1,125 @@
+     * bytes. This is the upper limit of {@link ByteBuf#readableBytes()}. If more than
+     * Integer#MAX_VALUE bytes are received a {@link TooMuchDataException} will be emitted.
+     * {@link TooMuchDataException#getCause()}
+     * will contain an
+     * {@link OnErrorThrowable.OnNextValue} with the bytes accumulated before the exception
+     * was thrown.
+     */
+    public static CollectBytes all() {
+        return upTo(Integer.MAX_VALUE);
+    }
+
+    /**
+     * Collect all emitted ByteBufs into a single ByteBuf until maxBytes have
+     * been collected. If more than maxBytes are received this will unsubscribe from
+     * the upstream Observable and will emit a
+     * {@link TooMuchDataException}. {@link TooMuchDataException#getCause()}
+     * will contain an
+     * {@link OnErrorThrowable.OnNextValue} with the bytes accumulated before the exception
+     * was thrown.
+     * @param maxBytes the maximum number of bytes to read
+     * @throws IllegalArgumentException when maxBytes is negative
+     */
+    public static CollectBytes upTo(int maxBytes) {
+        return new CollectBytes(maxBytes);
+    }
+
+    private CollectBytes(int maxBytes) {
+        if (maxBytes < 0) {
+            throw new IllegalArgumentException("maxBytes must not be negative");
+        }
+        this.maxBytes = maxBytes;
+    }
+
+    @Override
+    public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+        return upstream
+            .collect(
+                new Func0<CountingCollector>() {
+                    @Override
+                    public CountingCollector call() {
+                        return new CountingCollector();
+                    }
+                },
+                new Action2<CountingCollector, ByteBuf>() {
+                    @Override
+                    public void call(CountingCollector collector, ByteBuf buf) {
+                        collector.count += buf.readableBytes();
+                        if (collector.count <= maxBytes) {
+                            int i = buf.readableBytes();
+                            collector.byteBuf.addComponent(buf);
+                            collector.byteBuf.writerIndex(collector.byteBuf.writerIndex() + i);
+                        } else {
+                            collector.byteBuf.release();
+                            buf.release();
+                            throw new TooMuchDataException();
+                        }
+                    }
+                }
+            )
+            .map(new Func1<CountingCollector, ByteBuf>() {
+                @Override
+                public ByteBuf call(CountingCollector collector) {
+                    return collector.byteBuf;
+                }
+            });
+    }
+
+    public static class TooMuchDataException extends RuntimeException {
+        public TooMuchDataException() {
+            super();
+        }
+    }
+
+    private static class CountingCollector {
+        private long count;
```

Comment:
I think `CompositeBuffer.readableBytes()` would suffice instead of keeping this count, which in turn removes the need of this class.